### ISSUE TITLE
feature: Provide option to send no signal to exec process on reload

### ISSUE THREAD

**Issue by jjjordan** [hashicorp/consul-template#1428]
I would like there to be an option to not send a signal to the exec'd process when the configuration changes.  For reference, the documentation here states:

```hcl
  # This defines the signal that will be sent to the child process when a
  # change occurs in a watched template. The signal will only be sent after the
  # process is started, and the process will only be started after all
  # dependent templates have been rendered at least once. The default value is
  # nil, which tells Consul Template to stop the child process and spawn a new
  # one instead of sending it a signal. This is useful for legacy applications
  # or applications that cannot properly reload their configuration without a
  # full reload.
  reload_signal = ""
```

(Side note: Setting this to an empty string causes consul-template to exit with `* os: unsupported signal type`)

My reasoning here is this:
1. There is value in using exec beyond reloading the process. e.g., making sure all templates are rendered before starting the process that will consume them.
2. I can control reload through the `command` setting for the template, so long as there are other mechanisms to control the process (in my case, BIND9's `rndc`)
3. There is not always an *innocuous* signal available to choose.  Again for BIND, SIGHUP essentially restarts the server instead of just reloading the zone that was changed.  Any other signal kills it.
4. All of the workarounds I can come up with have drawbacks.
    * Generating templates `-once` then starting consul-template in the background requires something like `supervisord` to manage all the subprocesses, propagate signals, and reap zombies.
    * Exec'ing a shell wrapper to trap some innocuous signal (repeatedly) and pass on others is very tricky and error prone.

To be clear, this would be a nice-to-have and isn't something strictly necessary.  Perhaps assigning this behavior to an empty string would be confusing (since `nil` already has a meaning).  I'd considered a PR but as the signal is already passed as an `os.Signal` it would involve a lot of plumbing.

**Comment by eikenb** [comment]
Thanks, @jjjordan, for taking the time to file this issue.

I've labelled this a bug as I think they empty string should indicate no-signal as you first thought. It should not spit out an error, but the fact that it currently does is good as that means we can change its behavior without concerns about breaking current behavior.

**Comment by sdomagala** [comment]
Hi @eikenb, do you maybe have an update on this issue? I feel it's viable to have a possibility not to reload a process whenever the configuration changes but as far as I can tell by looking at the docs there is no such option now.

Thanks!

**Comment by eikenb** [comment]
Hey @sdomagala,

The best option is to set it to an unused signal, SIGUSR1 or SIGUSR2 are usually good candidates. Right now that is the only way to do it. 

Digging through the code a bit, I'm not sure if this is as hard as I first thought. It looks like we just need to handle the case of SIGNIL (signal's zero value) and skip sending it with that value. If I change the below code to just log the bad signal and return nil it works in my case...

https://github.com/hashicorp/consul-template/blob/78eedd3ba545cc4ce13e1e162735236291e70028/child/child.go#L362-L365

I'm sure there will be a bit more to it as that change breaks the tests (some fails, some hangs) but I think, with some work,  something like that will work.

Note that I'm not promising doing this right now, I'm in the middle of something else ATM, but I'll get to it.

**Comment by eikenb** [comment]
Added this one the next milestone so I won't forget it when I get back to working on another consul-template release. Sorry for the delay.

**Comment by eikenb** [comment]
PR up... https://github.com/hashicorp/consul-template/pull/1610
